File: Client/PyClient/Scripts/mesh_manager.py
#!/usr/bin/env python3
"""
Test script for adding and removing global meshes in VCCSim
"""

import os
import time
import sys
import struct
from typing import List, Dict, Tuple


sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from VCCSim import VCCSimClient


# Magic number expected by the UE mesh processor
MESH_MAGIC_NUMBER = 0x48534D55  # Hex representation of 'UMSH' in little-endian

def parse_obj_file(file_path: str) -> Tuple[List[Tuple[float, float, float]], List[List[int]]]:
    """
    Parse an OBJ file and extract vertices and triangles
    
    Args:
        file_path: Path to the OBJ file
        
    Returns:
        Tuple containing list of vertices and list of triangle indices
        Each triangle is a list of 3 indices
    """
    vertices = []
    triangles = []
    
    try:
        with open(file_path, 'r') as f:
            for line in f:
                parts = line.strip().split()
                if not parts:
                    continue
                
                if parts[0] == 'v':
                    # Vertex (convert to float)
                    v = (float(parts[1]), float(parts[2]), float(parts[3]))
                    vertices.append(v)
                elif parts[0] == 'f':
                    # Face (convert to 0-based indices)
                    # OBJ format uses 1-based indexing, so subtract 1
                    face_indices = []
                    for p in parts[1:]:
                        # Handle different face formats (v, v/vt, v/vt/vn, v//vn)
                        idx = p.split('/')[0]
                        if idx:
                            face_indices.append(int(idx) - 1)
                    
                    # Only add triangles
                    if len(face_indices) == 3:
                        triangles.append(face_indices)
                    elif len(face_indices) > 3:
                        # Simple triangulation of polygons (fan method)
                        for i in range(1, len(face_indices) - 1):
                            triangles.append([face_indices[0], face_indices[i], face_indices[i+1]])
        
        return vertices, triangles
    except Exception as e:
        print(f"Error parsing OBJ file {file_path}: {e}")
        sys.exit(1)

def create_mesh_binary(vertices: List[Tuple[float, float, float]], 
                      triangles: List[List[int]]) -> bytes:
    """
    Create a binary mesh representation that matches the expected format in UFMeshManager::ProcessMeshData
    
    FMeshHeader structure:
    struct FMeshHeader {
        uint32 Magic;       // Magic number
        uint32 Version;     // Version of the format
        uint32 VertexCount; // Number of vertices
        uint32 IndexCount;  // Number of indices
        uint32 Flags;       // Additional flags
    };
    
    FCompactVertex structure:
    struct FCompactVertex {
        float X, Y, Z;
    };
    
    Args:
        vertices: List of vertex positions (x, y, z)
        triangles: List of triangle indices, where each triangle is a list of 3 indices
        
    Returns:
        Binary data in the format expected by the server
    """
    # Calculate counts
    vertex_count = len(vertices)
    
    # Flatten triangle indices into a single list
    flat_indices = []
    for tri in triangles:
        flat_indices.extend(tri)
    index_count = len(flat_indices)
    
    # Define constants
    version = 1  # Version of the format
    flags = 0    # No special flags
    
    # Create header with all 5 fields (magic, version, vertex_count, index_count, flags)
    header = struct.pack('<IIIII', MESH_MAGIC_NUMBER, version, vertex_count, index_count, flags)
    
    # Create vertices block (FCompactVertex format: 3 floats)
    vertices_data = bytearray()
    for v in vertices:
        vertices_data.extend(struct.pack('<fff', v[0], v[1], v[2]))
    
    # Create indices block (uint32 format)
    indices_data = bytearray()
    for idx in flat_indices:
        indices_data.extend(struct.pack('<I', idx))
    
    # Combine into final binary mesh data
    mesh_data = header + vertices_data + indices_data
    
    # Debug info to verify sizes
    header_size = struct.calcsize('<IIIII')  # 5 uint32s
    vertex_size = struct.calcsize('<fff')    # 3 floats per vertex
    index_size = struct.calcsize('<I')       # 1 uint32 per index
    
    expected_size = header_size + (vertex_count * vertex_size) + (index_count * index_size)
    actual_size = len(mesh_data)
    
    print(f"Mesh binary size breakdown:")
    print(f"  Header: {header_size} bytes")
    print(f"  Vertices ({vertex_count}): {vertex_count * vertex_size} bytes")
    print(f"  Indices ({index_count}): {index_count * index_size} bytes")
    print(f"  Expected total: {expected_size} bytes")
    print(f"  Actual total: {actual_size} bytes")
    
    # Verify the binary size matches our calculation
    assert actual_size == expected_size, f"Size mismatch: expected {expected_size}, got {actual_size}"
    
    return mesh_data

def main():
    # Path to the OBJ files
    obj_dir = r"C:\UEProjects\VCCSimDev\Plugins\VCCSim\test\hull"
    
    # List all OBJ files
    obj_files = [os.path.join(obj_dir, f) for f in os.listdir(obj_dir) if f.endswith('.obj')]
    
    if not obj_files:
        print(f"No OBJ files found in {obj_dir}")
        sys.exit(1)
    
    print(f"Found {len(obj_files)} OBJ files:")
    for i, file_path in enumerate(obj_files):
        print(f"  {i+1}. {os.path.basename(file_path)}")
    
    # Initialize the client
    client = VCCSimClient()
    print("Connected to VCCSim server")
    
    # Dictionary to store mesh IDs
    mesh_ids: Dict[str, int] = {}
    
    try:
        # Add all meshes
        print("\nAdding global meshes...")
        for i, file_path in enumerate(obj_files):
            file_name = os.path.basename(file_path)
            
            # Parse OBJ file
            print(f"Parsing {file_name}...", end="", flush=True)
            vertices, triangles = parse_obj_file(file_path)
            print(f" Done. {len(vertices)} vertices, {len(triangles)} triangles")
            
            # Create binary mesh data
            mesh_data = create_mesh_binary(vertices, triangles)
            file_size = len(mesh_data) / 1024  # Size in KB
            
            # Different transform for each mesh to place them side by side
            x_offset = i * 5.0  # 5 units apart on X axis
            transform_pose = (x_offset, 0.0, 0.0, 0.0, 0.0, 0.0)  # x, y, z, roll, pitch, yaw
            
            # Send the mesh
            print(f"Sending {file_name} ({file_size:.2f} KB)...", end="", flush=True)
            mesh_id = client.send_global_mesh(
                data=mesh_data,
                format=0,          # Format identifier (adjust as needed)
                version=1,         # Format version (adjust as needed)
                simplified=False,  # Using full mesh detail
                transform_pose=transform_pose
            )
            
            mesh_ids[file_name] = mesh_id
            print(f" Done. Mesh ID: {mesh_id}")
            
            # Brief pause between adding meshes
            time.sleep(0.5)
        
        # Display all added meshes
        print("\nAdded global meshes:")
        for file_name, mesh_id in mesh_ids.items():
            print(f"  {file_name}: ID {mesh_id}")
        
        # Wait for user to observe the meshes in the simulator
        input("\nPress Enter to start removing meshes one by one...")
        
        # Remove meshes one by one
        print("\nRemoving global meshes...")
        for file_name, mesh_id in mesh_ids.items():
            print(f"Removing {file_name} (ID: {mesh_id})...", end="", flush=True)
            success = client.remove_global_mesh(mesh_id)
            print(f" {'Success' if success else 'Failed'}")
            
            # Brief pause between removals
            time.sleep(1.0)
        
        print("\nAll meshes removed.")
        
    except Exception as e:
        print(f"\nError during test: {e}")
    finally:
        # Clean up
        print("\nClosing connection...")
        client.close()
        print("Test completed.")

if __name__ == "__main__":
    main()

File: Client/CppClient/test_src/mesh_manager.cpp
/**
 * Test application for adding and removing global meshes in VCCSim
 * C++ version of the Python test script - Updated to match newest client API
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>
#include <tuple>
#include <chrono>
#include <thread>
#include <cstdint>
#include <sstream>  // Added missing include for istringstream
#include "VCCSimClient.h"

// For older MSVC compilers that don't use C++17 by default
#ifdef _MSC_VER
    #if _MSC_VER < 1914 || !defined(_MSVC_LANG) || _MSVC_LANG < 201703L
        // Define simple path-related functions to replace std::filesystem
        namespace fs_helper {
            bool exists(const std::string& path) {
                std::ifstream file(path);
                return file.good();
            }
            
            std::string filename(const std::string& path) {
                size_t pos = path.find_last_of("/\\");
                if (pos != std::string::npos) {
                    return path.substr(pos + 1);
                }
                return path;
            }
            
            std::string extension(const std::string& path) {
                size_t pos = path.find_last_of(".");
                if (pos != std::string::npos) {
                    return path.substr(pos);
                }
                return "";
            }
            
            std::vector<std::string> list_files(const std::string& dir, const std::string& ext) {
                std::vector<std::string> files;
                // This is a simplified version that works on Windows only
                // For a real project, use proper platform-specific code or third-party libraries
                std::string cmd = "dir /b \"" + dir + "\"";
                FILE* pipe = _popen(cmd.c_str(), "r");
                if (pipe) {
                    char buffer[256];
                    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
                        std::string file = buffer;
                        // Remove newline character
                        if (!file.empty() && file[file.size() - 1] == '\n') {
                            file.erase(file.size() - 1);
                        }
                        
                        if (extension(file) == ext) {
                            files.push_back(dir + "\\" + file);
                        }
                    }
                    _pclose(pipe);
                }
                return files;
            }
        }
    #else
        #include <filesystem>
        namespace fs_helper = std::filesystem;
    #endif
#else
    #include <filesystem>
    namespace fs_helper = std::filesystem;
#endif

// Magic number expected by the UE mesh processor - 'UMSH' in little-endian
const uint32_t MESH_MAGIC_NUMBER = 0x48534D55;

// Struct to represent a 3D vertex
struct Vertex {
    float x, y, z;
};

// Typedef for triangle indices
typedef std::vector<uint32_t> Triangle;

// Function to parse an OBJ file and extract vertices and triangles
std::pair<std::vector<Vertex>, std::vector<Triangle>> parseObjFile(const std::string& filePath) {
    std::vector<Vertex> vertices;
    std::vector<Triangle> triangles;
    
    std::ifstream file(filePath);
    if (!file.is_open()) {
        std::cerr << "Error opening OBJ file: " << filePath << std::endl;
        exit(1);
    }
    
    std::string line;
    while (std::getline(file, line)) {
        if (line.empty()) continue;
        
        std::istringstream iss(line);
        std::string token;
        iss >> token;
        
        if (token == "v") {
            // Parse vertex
            Vertex v;
            if (iss >> v.x >> v.y >> v.z) {
                vertices.push_back(v);
            }
        }
        else if (token == "f") {
            // Parse face
            std::vector<uint32_t> faceIndices;
            std::string indexStr;
            
            while (iss >> indexStr) {
                // Handle different face formats (v, v/vt, v/vt/vn, v//vn)
                size_t slashPos = indexStr.find('/');
                std::string vertexIndexStr;
                
                if (slashPos != std::string::npos) {
                    vertexIndexStr = indexStr.substr(0, slashPos);
                } else {
                    vertexIndexStr = indexStr;
                }
                
                // OBJ format uses 1-based indexing, so subtract 1
                if (!vertexIndexStr.empty()) {
                    uint32_t index = std::stoi(vertexIndexStr) - 1;
                    faceIndices.push_back(index);
                }
            }
            
            // Only add triangles, or triangulate if more than 3 vertices
            if (faceIndices.size() == 3) {
                triangles.push_back({faceIndices[0], faceIndices[1], faceIndices[2]});
            }
            else if (faceIndices.size() > 3) {
                // Simple triangulation of polygons (fan method)
                for (size_t i = 1; i < faceIndices.size() - 1; ++i) {
                    triangles.push_back({faceIndices[0], faceIndices[i], faceIndices[i+1]});
                }
            }
        }
    }
    
    file.close();
    return std::make_pair(vertices, triangles);
}

// Function to create binary mesh data in the format expected by UFMeshManager::ProcessMeshData
std::vector<uint8_t> createMeshBinary(const std::vector<Vertex>& vertices, 
                                     const std::vector<Triangle>& triangles) {
    // Calculate counts
    uint32_t vertexCount = static_cast<uint32_t>(vertices.size());
    
    // Flatten triangle indices into a single vector
    std::vector<uint32_t> flatIndices;
    for (const auto& tri : triangles) {
        flatIndices.insert(flatIndices.end(), tri.begin(), tri.end());
    }
    uint32_t indexCount = static_cast<uint32_t>(flatIndices.size());
    
    // Define constants
    uint32_t version = 1;  // Version of the format
    uint32_t flags = 0;    // No special flags
    
    // Calculate sizes
    const size_t headerSize = 5 * sizeof(uint32_t);  // 5 uint32 fields
    const size_t vertexSize = 3 * sizeof(float);     // 3 floats per vertex
    const size_t indexSize = sizeof(uint32_t);       // 1 uint32 per index
    size_t totalSize = headerSize + (vertexCount * vertexSize) + (indexCount * indexSize);
    
    // Prepare the binary data buffer
    std::vector<uint8_t> meshData(totalSize);
    uint8_t* buffer = meshData.data();
    size_t offset = 0;
    
    // Write header
    memcpy(buffer + offset, &MESH_MAGIC_NUMBER, sizeof(uint32_t));
    offset += sizeof(uint32_t);
    memcpy(buffer + offset, &version, sizeof(uint32_t));
    offset += sizeof(uint32_t);
    memcpy(buffer + offset, &vertexCount, sizeof(uint32_t));
    offset += sizeof(uint32_t);
    memcpy(buffer + offset, &indexCount, sizeof(uint32_t));
    offset += sizeof(uint32_t);
    memcpy(buffer + offset, &flags, sizeof(uint32_t));
    offset += sizeof(uint32_t);
    
    // Write vertices
    for (const auto& v : vertices) {
        memcpy(buffer + offset, &v.x, sizeof(float));
        offset += sizeof(float);
        memcpy(buffer + offset, &v.y, sizeof(float));
        offset += sizeof(float);
        memcpy(buffer + offset, &v.z, sizeof(float));
        offset += sizeof(float);
    }
    
    // Write indices
    for (const auto& idx : flatIndices) {
        memcpy(buffer + offset, &idx, sizeof(uint32_t));
        offset += sizeof(uint32_t);
    }
    
    // Debug info to verify sizes
    std::cout << "Mesh binary size breakdown:" << std::endl;
    std::cout << "  Header: " << headerSize << " bytes" << std::endl;
    std::cout << "  Vertices (" << vertexCount << "): " << (vertexCount * vertexSize) << " bytes" << std::endl;
    std::cout << "  Indices (" << indexCount << "): " << (indexCount * indexSize) << " bytes" << std::endl;
    std::cout << "  Expected total: " << totalSize << " bytes" << std::endl;
    std::cout << "  Actual total: " << meshData.size() << " bytes" << std::endl;
    
    // Verify the binary size matches our calculation
    if (meshData.size() != totalSize) {
        std::cerr << "Size mismatch: expected " << totalSize << ", got " << meshData.size() << std::endl;
        exit(1);
    }
    
    return meshData;
}

int main() {
    // Path to the OBJ files
    std::string objDir = "C:\\UEProjects\\VCCSimDev\\Plugins\\VCCSim\\test\\hull";
    
    // List all OBJ files
    std::vector<std::string> objFiles;
    
    try {
        #ifdef _MSC_VER
            #if _MSC_VER < 1914 || !defined(_MSVC_LANG) || _MSVC_LANG < 201703L
                // Use the helper for older compilers
                objFiles = fs_helper::list_files(objDir, ".obj");
            #else
                // Use std::filesystem for C++17 support
                for (const auto& entry : fs_helper::directory_iterator(objDir)) {
                    if (fs_helper::is_regular_file(entry) && fs_helper::path(entry).extension() == ".obj") {
                        objFiles.push_back(entry.path().string());
                    }
                }
            #endif
        #else
            // Use std::filesystem for non-MSVC compilers
            for (const auto& entry : fs_helper::directory_iterator(objDir)) {
                if (fs_helper::is_regular_file(entry) && entry.path().extension() == ".obj") {
                    objFiles.push_back(entry.path().string());
                }
            }
        #endif
    } catch (const std::exception& e) {
        std::cerr << "Error listing OBJ files: " << e.what() << std::endl;
        return 1;
    }
    
    if (objFiles.empty()) {
        std::cerr << "No OBJ files found in " << objDir << std::endl;
        return 1;
    }
    
    std::cout << "Found " << objFiles.size() << " OBJ files:" << std::endl;
    for (size_t i = 0; i < objFiles.size(); ++i) {
        #ifdef _MSC_VER
            #if _MSC_VER < 1914 || !defined(_MSVC_LANG) || _MSVC_LANG < 201703L
                std::cout << "  " << (i+1) << ". " << fs_helper::filename(objFiles[i]) << std::endl;
            #else
                std::cout << "  " << (i+1) << ". " << fs_helper::path(objFiles[i]).filename().string() << std::endl;
            #endif
        #else
            std::cout << "  " << (i+1) << ". " << fs_helper::path(objFiles[i]).filename().string() << std::endl;
        #endif
    }
    
    // Initialize the client
    VCCSimClient client;
    std::cout << "Connected to VCCSim server" << std::endl;
    
    // Map to store mesh IDs
    std::map<std::string, int> meshIds;
    
    try {
        // Add all meshes
        std::cout << "\nAdding global meshes..." << std::endl;
        for (size_t i = 0; i < objFiles.size(); ++i) {
            std::string filePath = objFiles[i];
            
            // Get file name for display
            std::string fileName;
            #ifdef _MSC_VER
                #if _MSC_VER < 1914 || !defined(_MSVC_LANG) || _MSVC_LANG < 201703L
                    fileName = fs_helper::filename(filePath);
                #else
                    fileName = fs_helper::path(filePath).filename().string();
                #endif
            #else
                fileName = fs_helper::path(filePath).filename().string();
            #endif
            
            // Parse OBJ file
            std::cout << "Parsing " << fileName << "..." << std::flush;
            auto parsed = parseObjFile(filePath);
            const auto& vertices = parsed.first;
            const auto& triangles = parsed.second;
            std::cout << " Done. " << vertices.size() << " vertices, " << triangles.size() << " triangles" << std::endl;
            
            // Create binary mesh data
            std::vector<uint8_t> meshData = createMeshBinary(vertices, triangles);
            float fileSize = meshData.size() / 1024.0f;  // Size in KB
            
            // Convert binary data to string
            std::string meshDataStr(reinterpret_cast<char*>(meshData.data()), meshData.size());
            
            // Different transform for each mesh to place them side by side
            float xOffset = i * 5.0f;  // 5 units apart on X axis
            
            // Create a tuple for the transform pose (x, y, z, roll, pitch, yaw)
            auto transformPose = std::make_tuple(xOffset, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f);
            
            // Send the mesh using the updated API
            std::cout << "Sending " << fileName << " (" << fileSize << " KB)..." << std::flush;
            int meshId = client.SendGlobalMesh(
                meshDataStr,          // String data instead of raw bytes
                0,                    // Format identifier
                1,                    // Format version
                false,                // Using full mesh detail
                transformPose         // Transform pose as a tuple
            );
            
            meshIds[fileName] = meshId;
            std::cout << " Done. Mesh ID: " << meshId << std::endl;
            
            // Brief pause between adding meshes
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
        
        // Display all added meshes
        std::cout << "\nAdded global meshes:" << std::endl;
        for (const auto& pair : meshIds) {
            const auto& fileName = pair.first;
            const auto& meshId = pair.second;
            std::cout << "  " << fileName << ": ID " << meshId << std::endl;
        }
        
        // Wait for user to observe the meshes in the simulator
        std::cout << "\nPress Enter to start removing meshes one by one..." << std::endl;
        std::cin.get();
        
        // Remove meshes one by one
        std::cout << "\nRemoving global meshes..." << std::endl;
        for (const auto& pair : meshIds) {
            const auto& fileName = pair.first;
            const auto& meshId = pair.second;
            std::cout << "Removing " << fileName << " (ID: " << meshId << ")..." << std::flush;
            bool success = client.RemoveGlobalMesh(meshId);
            std::cout << " " << (success ? "Success" : "Failed") << std::endl;
            
            // Brief pause between removals
            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
        }
        
        std::cout << "\nAll meshes removed." << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "\nError during test: " << e.what() << std::endl;
    }
    
    // Cleanup happens in the VCCSimClient destructor
    std::cout << "\nClosing connection..." << std::endl;
    std::cout << "Test completed." << std::endl;
    
    return 0;
}